### ISSUE TITLE
Validator: Account not found: key not found rpc error

### ISSUE THREAD

**Issue by prabhanjangururaj** [provenance-io/provenance#459]
## Summary of Bug

I hit an rpc error as key not found as below while creating a validator node on my ubuntu machine. 

```
$ provenanced tx staking create-validator --moniker secured_provenance --pubkey tpvalconspub1zcjduepq5x4s983f2ueq3tr6mnkzcc3jvp637nnmu4swu4727c6u959qw9vqua29kg  --commission-rate=1.0 --commission-max-rate=1.0 --commission-max-change-rate=1.0 --min-self-delegation=1 --from=secure_key --amount 65000nhash --keyring-dir /home/ubuntu/.config/Provenance/ -t --broadcast-mode block
Error: rpc error: code = NotFound desc = account tp14qrnd9jplk4ag2e8620396mrag34esvenlaj6s not found: key not found
```

I have used the pub key derived from the show-validator command as below:
```
$ provenanced --testnet tendermint show-validator
tpvalconspub1zcjduepq5x4s983f2ueq3tr6mnkzcc3jvp637nnmu4swu4727c6u959qw9vqua29kg
```

The account and key-dir seems to be valid and active:
```
$ provenanced --testnet q bank balances tp14qrnd9jplk4ag2e8620396mrag34esvenlaj6s --node=tcp://rpc-0.test.provenance.io:26657
balances:
- amount: "2000000000"
  denom: nhash
pagination:
  next_key: null
  total: "0"
```

```
$ provenanced keys list --keyring-backend test -t
- name: secure_key
  type: local
  address: tp14qrnd9jplk4ag2e8620396mrag34esvenlaj6s
  pubkey: tppub1addwnpepqwgpkqemkvxjwv2nuvz5jehu9z89qcx5vme8356q5klh3lrj3ds2xw4lqg5
  mnemonic: ""
  threshold: 0
  pubkeys: []
```

## Version

$ provenanced version
v1.0.0

## Steps to Reproduce

```
git clone -b v1.0.0 https://github.com/provenance-io/provenance
cd provenance && make install
export PATH=$PATH:/home/ubuntu/go/bin
provenanced init secured_provenance --testnet
curl https://raw.githubusercontent.com/provenance-io/testnet/main/pio-testnet-1/genesis.json > genesis.json
export PIO_HOME=/home/ubuntu/.config/Provenance
mkdir -p $PIO_HOME/config
mv genesis.json $PIO_HOME/config
Edit $PIO_HOME/config/config.toml with 
    db_backend = "cleveldb" 
	seeds = "2de841ce706e9b8cdff9af4f137e52a4de0a85b2@104.196.26.176:26656,add1d50d00c8ff79a6f7b9873cc0d9d20622614e@34.71.242.51"
	namespace = "provenance"
go get github.com/provenance-io/cosmovisor/cmd/cosmovisor    
mkdir -p $PIO_HOME/cosmovisor/install
cd $PIO_HOME/cosmovisor/install
git clone https://github.com/provenance-io/cosmovisor.git
cd cosmovisor
make
cp build/cosmovisor /home/ubuntu/go/bin/cosmovisor

export DAEMON_NAME="provenanced"
export DAEMON_HOME="${PIO_HOME}"
export DAEMON_ALLOW_DOWNLOAD_BINARIES="true" 
export DAEMON_RESTART_AFTER_UPGRADE="true"

# Set up keys
# Whilelist the account 


mkdir -p $PIO_HOME/cosmovisor/genesis/bin
mkdir -p $PIO_HOME/cosmovisor/upgrades
ln -sf $PIO_HOME/cosmovisor/genesis/bin $PIO_HOME/cosmovisor/genesis/current

cp $(which provenanced) $PIO_HOME/cosmovisor/genesis/bin 
ln -sf $PIO_HOME/cosmovisor/genesis/bin/provenanced $(which provenanced)

# get the pub key 
provenanced --testnet tendermint show-validator   
		
# Run full node
cosmovisor start --testnet --home $PIO_HOME --p2p.seeds 2de841ce706e9b8cdff9af4f137e52a4de0a85b2@104.196.26.176:26656,add1d50d00c8ff79a6f7b9873cc0d9d20622614e@34.71.242.51:26656 --x-crisis-skip-assert-invariants

# Run validator
provenanced  tx staking create-validator --moniker secured_provenance --pubkey tpvalconspub1zcjduepq5x4s983f2ueq3tr6mnkzcc3jvp637nnmu4swu4727c6u959qw9vqua29kg  --commission-rate=1.0 --commission-max-rate=1.0 --commission-max-change-rate=1.0 --min-self-delegation=1 --from=secure_key --amount 65000nhash --keyring-dir /home/ubuntu/.config/Provenance/ -t --broadcast-mode block
```

____

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by iramiller** [comment]
You have a potentially different paths on the keys list command and the others. For consistency use the same `--home` folder flag on all commands to ensure you are working with the same folder in each case. 

The error you are getting simply means your keyring does not contain the key listed.

**Comment by prabhanjangururaj** [comment]
It all looks good there Ira Miller. I tried again and hitting the same issue. What else can I verify?

Tried with an earlier version this time -
$ provenanced version
0.2.0

**Comment by iramiller** [comment]
I'm out for a few days... maybe @arnabmitra can assist or pull someone in to help you.

**Comment by arnabmitra** [comment]
> It all looks good there Ira Miller. I tried again and hitting the same issue. What else can I verify?
> 
> Tried with an earlier version this time -
> $ provenanced version
> 0.2.0

Hi, 
the current provenance version is 1.6.0 on testnet, however i think your problem is that the key is not being found in you create validator command
1. remove ` --keyring-dir `
2.  add `--keyring-backend test` to your existing cmd

not specifying a key ring backend default it to `os` which you are not setting up from what i can see
https://docs.cosmos.network/v0.43/run-node/keyring.html for more info.

**Comment by channa-figure** [comment]
Also, have you tried any other transactions with that account?  It would be a good test to try and send some `nhash` with a bank tx.  

Example: 
`provenanced tx bank send tp14qrnd9jplk4ag2e8620396mrag34esvenlaj6s <to_address> 100000000nhash --fees 1000000000nhash --chain-id pio-testnet-1 --keyring-backend test --testnet --home $PIO_HOME`

**Comment by prabhanjangururaj** [comment]
Thanks for your response Arnab and Carlton.

I can see the key when I specify  `--keyring-backend test` and also used the same in my command to start a validator or send some nhash. Both failed again with the same `key not found` issue. Where querying balance with my address still works as I had shown earlier. 

```
$ provenanced keys show secure_key -t --keyring-backend test
- name: secure_key
  type: local
  address: tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn
  pubkey: tppub1addwnpepq2k7rx5huv73r8ap4h3umk7a206uwv0x6lhv705tddev5dwk89t6w2wds54
  mnemonic: ""
  threshold: 0
  pubkeys: []
```

Noting my validator pub key:
```
$ provenanced --testnet tendermint show-validator --home $PIO_HOME
tpvalconspub1zcjduepqkrgu9k0p8dkvcma2uslhmpu6mdhf87etykwlfk456duy0hxtwqjq288jt3
```

Now both creating a validator or sending nhash fails even after specifying keyring-backend as test:
```
$ provenanced --testnet tx bank send  tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn tp1qhutk7wuj96ufq96v2musp4gpqgphpzk37nq94 100000000nhash --fees 1000000000nhash --chain-id pio-testnet-1 --keyring-backend test --testnet --home $PIO_HOME
Error: rpc error: code = NotFound desc = account tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn not found: key not found

$ provenanced  tx staking create-validator --moniker provenanced-node --pubkey tpvalconspub1zcjduepqkrgu9k0p8dkvcma2uslhmpu6mdhf87etykwlfk456duy0hxtwqjq288jt3  --commission-rate=1.0 --commission-max-rate=1.0 --commission-max-change-rate=1.0 --min-self-delegation=1 --from=secure_key --amount 65000nhash  -t --broadcast-mode block --keyring-backend test
Error: rpc error: code = NotFound desc = account tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn not found: key not found
```

While I continue to try and debug this with your help, do you suggest me any alternate ways to create a validator? May be a container image which I can try and run as a validator in testnet network?

Thanks.

**Comment by arnabmitra** [comment]
Hi @prabhanjangururaj , what is the result of the cmd `provenanced keys show secure_key -t --keyring-backend test  --home $PIO_HOME`, imo it will give you an  error 

note this(from my mac may differ on your box)

 ```Global Flags:
      --home string              The application home directory (default "/Users/xxx/.provenanced")
      --keyring-backend string   Select keyring's backend (os|file|test) (default "test")
      --keyring-dir string       The client Keyring directory; if omitted, the default 'home' directory will be used```

tl;dr if you are defining home define the same when you do the key import `provenanced keys import name --home $PIO_HOME`

i.e when you import the keys you cannot say goto default `home` and then ask it to look at specific --home you specify


you can find the docker images here https://hub.docker.com/r/provenanceio/provenance

**Comment by scirner22** [comment]
These guys can provide more help than I can, but testnet would take quite a long time to sync and catch up to the head state. It would be worthwhile to try your `tx bank send` attempt again with `--node=tcp://rpc-0.test.provenance.io:26657` added just so you know the request is going to a node that's caught up and not behind at all.

I'm assuming here that if you submit to a local node that hasn't replicated to the block where that address existed and received hash, it would return the error you're seeing. This assumption could be totally wrong.

**Comment by prabhanjangururaj** [comment]
> provenanced keys import name --home $PIO_HOME

Hi Arnab - this is the result of the command:

```
$ provenanced keys show secure_key -t --keyring-backend test --home $PIO_HOME
- name: secure_key
  type: local
  address: tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn
  pubkey: tppub1addwnpepq2k7rx5huv73r8ap4h3umk7a206uwv0x6lhv705tddev5dwk89t6w2wds54
  mnemonic: ""
  threshold: 0
  pubkeys: []
```

Do you see any problem here?

**Comment by prabhanjangururaj** [comment]
> --node=tcp://rpc-0.test.provenance.io:26657

Hi Stephen - 

That made sense and here is what I observe now with `tx send`.  Thanks

```
$ provenanced --testnet tx bank send  tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn tp1qhutk7wuj96ufq96v2musp4gpqgphpzk37nq94 100000000nhash --fees 1000000000nhash --chain-id pio-testnet-1 --keyring-backend test --testnet --home $PIO_HOME --node=tcp://rpc-0.test.provenance.io:26657
{"body":{"messages":[{"@type":"/cosmos.bank.v1beta1.MsgSend","from_address":"tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn","to_address":"tp1qhutk7wuj96ufq96v2musp4gpqgphpzk37nq94","amount":[{"denom":"nhash","amount":"100000000"}]}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[{"denom":"nhash","amount":"1000000000"}],"gas_limit":"200000","payer":"","granter":""}},"signatures":[]}

confirm transaction before signing and broadcasting [y/N]: y
{"height":"3367587","txhash":"3AC979A42DE93BDD407F848F5484646BDAFFD4FCD7AB7618D826429E94821790","codespace":"","code":0,"data":"0A060A0473656E64","raw_log":"[{\"events\":[{\"type\":\"message\",\"attributes\":[{\"key\":\"action\",\"value\":\"send\"},{\"key\":\"sender\",\"value\":\"tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn\"},{\"key\":\"module\",\"value\":\"bank\"}]},{\"type\":\"transfer\",\"attributes\":[{\"key\":\"recipient\",\"value\":\"tp1qhutk7wuj96ufq96v2musp4gpqgphpzk37nq94\"},{\"key\":\"sender\",\"value\":\"tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn\"},{\"key\":\"amount\",\"value\":\"100000000nhash\"}]}]}]","logs":[{"msg_index":0,"log":"","events":[{"type":"message","attributes":[{"key":"action","value":"send"},{"key":"sender","value":"tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn"},{"key":"module","value":"bank"}]},{"type":"transfer","attributes":[{"key":"recipient","value":"tp1qhutk7wuj96ufq96v2musp4gpqgphpzk37nq94"},{"key":"sender","value":"tp1y7yhjnegfy86pctmva8helpf96dhua73fxq8pn"},{"key":"amount","value":"100000000nhash"}]}]}],"info":"","gas_wanted":"200000","gas_used":"154691","tx":null,"timestamp":""}
```

**Comment by prabhanjangururaj** [comment]
I wanted to have the sync complete and evaluate further, but the process crashed as below:

```
5:28AM INF ABCI Handshake App Info hash="0�;�1D\x18�\x03�O^�2\x02\u007f/\x19�t��S�\x15��\f���i" height=609072 module=consensus protocol-version=0 software-version=v1.7.0
5:28AM INF ABCI Replay Blocks appHeight=609072 module=consensus stateHeight=609072 storeHeight=609073
5:28AM INF Replay last block using real app module=consensus
Error: error during handshake: error on replay: wrong Block.Header.LastResultsHash.  Expected 567F1092207C8A52AA2E6CC4D697C4167B8815FABF07B9C88B6F2B7F9B1BD9F5, got 50F1D9F2DD2CD7971A4AFA9DF5D3B554C3050988247C9175994ADF131C69B182
```
I am right now on the latest version:
```
$ provenanced version
v1.7.0
```

**Comment by iramiller** [comment]
Version of provenanced must match the version at a given blockheight. Neither testnet nor mainnet are running 1.7.0 at this time. If you are syncing from zero start with the genesis version and upgrade as instructed by the software at the required blockheights

**Comment by prabhanjangururaj** [comment]
Hi Ira - I moved to 1.7.0 hoping to resolve this, but hit this issue even at 1.0.0. Can you suggest ways to recover from here and continue syncing? As you had pointed, I started with 0.2.0, the genesis version initially.

**Comment by iramiller** [comment]
You must match the version used when the block was cut... meaning you will need to transition through each version as your sync reaches the upgrade heights. 1.7.0 is not currently used in any public network (although testnet has an upgrade schedule to start minting blocks with it tomorrow). 

Carefully review your output log for the upgrade needed statement which is output at the upgrade height. Alternatively use the cosmovisor setup (on our fork as the upstream version is broken) and it will help you manage these upgrades.

**Comment by prabhanjangururaj** [comment]
Hi Ira - I will initiate again as suggested. 

I find a container to create a node on testnet here https://github.com/provenance-io/testnet/tree/main/docker. Have been running this without issues syncing close to a million blocks in the last 30hrs or so. 
```
$ provenanced  --testnet q block | jq ".block.last_commit.height"
"939263"
```

**Comment by prabhanjangururaj** [comment]
The promotion to a validator node did happen now. Thanks all of you for your assistance so far: 
https://explorer.test.provenance.io/validator/tp13f24k3zauhackjs6dqsj2zyvnf7zshmx32c0gq
